pdp/handlers.go

```diff
@@ -0,0 +1,1020 @@
+			return
+		}
+
+		for _, subrootEntry := range addRootReq.Subroots {
+			if subrootEntry.SubrootCID == "" {
+				http.Error(w, "subrootCid is required for each subroot", http.StatusBadRequest)
+				return
+			}
+			if _, exists := subrootCIDsSet[subrootEntry.SubrootCID]; exists {
+				http.Error(w, "duplicate subrootCid in request", http.StatusBadRequest)
+				return
+			}
+
+			subrootCIDsSet[subrootEntry.SubrootCID] = struct{}{}
+		}
+	}
+
+	// Convert set to slice
+	subrootCIDsList := make([]string, 0, len(subrootCIDsSet))
+	for cidStr := range subrootCIDsSet {
+		subrootCIDsList = append(subrootCIDsList, cidStr)
+	}
+
+	// Map to store subrootCID -> [pieceInfo, pdp_pieceref.id, subrootOffset]
+	type SubrootInfo struct {
+		PieceInfo     abi.PieceInfo
+		PDPPieceRefID int64
+		SubrootOffset uint64
+	}
+
+	subrootInfoMap := make(map[string]*SubrootInfo)
+
+	// Start a DB transaction
+	_, err = p.db.BeginTransaction(ctx, func(tx *harmonydb.Tx) (bool, error) {
+		// Step 4: Get pdp_piecerefs matching all subroot cids + make sure those refs belong to serviceLabel
+		rows, err := tx.Query(`
+            SELECT ppr.piece_cid, ppr.id AS pdp_pieceref_id, ppr.piece_ref,
+                   pp.piece_padded_size
+            FROM pdp_piecerefs ppr
+            JOIN parked_piece_refs pprf ON pprf.ref_id = ppr.piece_ref
+            JOIN parked_pieces pp ON pp.id = pprf.piece_id
+            WHERE ppr.service = $1 AND ppr.piece_cid = ANY($2)
+        `, serviceLabel, subrootCIDsList)
+		if err != nil {
+			return false, err
+		}
+		defer rows.Close()
+
+		foundSubroots := make(map[string]struct{})
+		for rows.Next() {
+			var pieceCIDStr string
+			var pdpPieceRefID, pieceRefID int64
+			var piecePaddedSize uint64
+
+			err := rows.Scan(&pieceCIDStr, &pdpPieceRefID, &pieceRefID, &piecePaddedSize)
+			if err != nil {
+				return false, err
+			}
+
+			// Parse the piece CID
+			pieceCID, err := cid.Decode(pieceCIDStr)
+			if err != nil {
+				return false, fmt.Errorf("invalid piece CID in database: %s", pieceCIDStr)
+			}
+
+			// Create PieceInfo
+			pieceInfo := abi.PieceInfo{
+				Size:     abi.PaddedPieceSize(piecePaddedSize),
+				PieceCID: pieceCID,
+			}
+
+			subrootInfoMap[pieceCIDStr] = &SubrootInfo{
+				PieceInfo:     pieceInfo,
+				PDPPieceRefID: pdpPieceRefID,
+				SubrootOffset: 0, // Will compute offset later
+			}
+
+			foundSubroots[pieceCIDStr] = struct{}{}
+		}
+
+		// Check if all subroot CIDs were found
+		for _, cidStr := range subrootCIDsList {
+			if _, found := foundSubroots[cidStr]; !found {
+				return false, fmt.Errorf("subroot CID %s not found or does not belong to service %s", cidStr, serviceLabel)
+			}
+		}
+
+		// Now, for each AddRootRequest, validate RootCID and prepare data for ETH transaction
+		for _, addRootReq := range req {
+			// Collect pieceInfos for subroots
+			pieceInfos := make([]abi.PieceInfo, len(addRootReq.Subroots))
+
+			var totalOffset uint64 = 0
```

Comment:
am I correct that reading this code right here provides indirect instructions on how to aggregate all the subroot pieces into a root CID? IOW, I can assume that I can basically want to write aggregation code that will satisfy this code by generating the same RootCID? (which appears to be all the pieces concatenated together). 

Also it would appear non-ffi is a fork of go-commp-utils -- is this public?

Reply:
Yes, the subroot aggregation is the same as in filecoin pieces (CommP -> CommD)

The non-ffi thing is the same code as the Rust logic, reimplemented by Riba in Go.

There is one added constraint in PDP - pieces must be sorted by size in descending order - this way no inter-piece padding is needed

Reply:
ooh good call out good to know!